File: client/Services/Connection.cs
using System;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using client.Models;
using client.Helpers;

namespace client.Services
{
    public class ConnectionService
    {
        private readonly HttpClient _client;
        private readonly IpService _ipService;
        private readonly JsonSerializerOptions _jsonOptions;

        public ConnectionService()
        {
            _client = new HttpClient();
            _ipService = new IpService();
            _jsonOptions = new JsonSerializerOptions
            {
                PropertyNamingPolicy = null,
                WriteIndented = true
            };
        }

        public async Task<ConnectionResponse> ConnectAsync()
        {
            try 
            {
                var ip = await _ipService.GetPublicIpAsync();
                var deviceInfo = $"{Environment.OSVersion}";
                
                var connectionInfo = new ConnectionRequest
                {
                    Ip = ip,
                    DeviceInfo = deviceInfo
                };

                var jsonRequest = JsonSerializer.Serialize(connectionInfo, _jsonOptions);
                // Console.WriteLine($"Sending JSON to server: {jsonRequest}");
                // Console.WriteLine($"Target URL: {Constants.API_BASE}/connections");
                
                var content = new StringContent(jsonRequest, Encoding.UTF8, "application/json");
                
                // Console.WriteLine("Request Headers:");
                // Console.WriteLine($"Content-Type: {content.Headers.ContentType}");
                
                var response = await _client.PostAsync($"{Constants.API_BASE}/connections", content);
                
                // Console.WriteLine($"Response Status: {response.StatusCode}");
                // Console.WriteLine("Response Headers:");
                foreach (var header in response.Headers)
                {
                    // Console.WriteLine($"{header.Key}: {string.Join(", ", header.Value)}");
                }
                
                if (!response.IsSuccessStatusCode)
                {
                    var errorContent = await response.Content.ReadAsStringAsync();
                    throw new HttpRequestException($"Server returned {response.StatusCode}: {errorContent}");
                }

                var responseBody = await response.Content.ReadAsStringAsync();
                // Console.WriteLine($"Server response: {responseBody}");
                
                return JsonSerializer.Deserialize<ConnectionResponse>(responseBody, _jsonOptions);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in ConnectAsync: {ex.Message}");
                if (ex.InnerException != null)
                {
                    Console.WriteLine($"Inner Exception: {ex.InnerException.Message}");
                }
                throw;
            }
        }

        public async Task DisconnectAsync(string connectionId)
        {
            if (string.IsNullOrEmpty(connectionId))
                throw new ArgumentException("Connection ID cannot be null or empty", nameof(connectionId));

            var response = await _client.PutAsync($"{Constants.API_BASE}/connections/{connectionId}/inactive", null);
            if (!response.IsSuccessStatusCode)
            {
                throw new HttpRequestException($"Failed to disconnect: {response.StatusCode}");
            }
        }
    }
}

File: client/Services/Connection.cpp
#include "connection.h"
#include "../network/http_client.h"
#include "../helpers/constants.h"
#include "../helpers/system_info.h"
#include "../helpers/parser.h"
#include <iostream>
#include <stdexcept>

ConnectionResponse ConnectionService::Connect() {
    auto ip = _ipService.GetPublicIp();
    auto deviceInfo = SystemInfo::GetDeviceInfo();

    ConnectionRequest request;
    request.ip = ip;
    request.deviceInfo = deviceInfo;
    request.ownerId = Constants::OWNER_ID;

    std::string url = Constants::API_BASE + "/connections";
    auto response = HttpClient::Post(url, request.toJson());

    if (response.statusCode < 200 || response.statusCode >= 300) {
        throw std::runtime_error("Server returned " + std::to_string(response.statusCode) + ": " + response.body);
    }

    auto json = JsonParser::Parse(response.body);

    ConnectionResponse result;
    result.id = json["_id"];
    result.ip = json["ip"];
    result.deviceInfo = json["deviceInfo"];
    result.timestamp = json["timestamp"];
    result.isActive = (json["isActive"] == "true");

    _heartbeat = std::make_unique<Heartbeat>(result.id);

    return result;
}

void ConnectionService::Disconnect(const std::string& connectionId) {
    if (connectionId.empty()) {
        throw std::invalid_argument("connectionId cannot be empty");
    }

    if (_heartbeat) {
        _heartbeat->Stop();
        _heartbeat.reset();
    }

    std::string url = Constants::API_BASE + "/connections/" + connectionId + "/inactive";
    auto response = HttpClient::Put(url);

    if (response.statusCode < 200 || response.statusCode >= 300) {
        throw std::runtime_error("Failed to disconnect: " + std::to_string(response.statusCode));
    }
}
